### DISCUSSION TITLE
Grafana EE - MongoDB UI/UX improvement suggestions

### DISCUSSION THREAD

**Discussion by calixte-mayoraz-eversys** [grafana/grafana#53508] [Data Sources]
Hello, first off let me start off by saying that the MongoDB plugin in Grafana EE has been an absolute game changer for us. However, we meet a few road-blocks which can really slow down efficient query-writing, and here are a few suggestions we have to improve it.

## 1) Make the query panel resizeable (or at least bigger...)

We write a lot of very long aggregation pipelines to get the visualizations we want. And editing these pipelines in a 150px tall textbox is quite a challenge. Most of the time, I cut/paste in Notepad++ edit it there, and then copy/paste it back in Grafana.

Having a larger text box, or (like for the SQL queries) having a text box which expands to fit the query would be nice. Or just something that we could resize in the bottom right (like a HTML textarea) would be the best

> before
>![image](https://user-images.githubusercontent.com/110970317/183855096-0a88f37f-ca0e-44ea-b253-6295e1ea1ab6.png)

> after (with a little tweaking of the page's CSS)
> ![image](https://user-images.githubusercontent.com/110970317/183855176-e95723b3-f827-4730-9f4d-c650951deab1.png)

## 2) Cleaner Query Inspector
Currently, when we need to debug our queries and inspect the raw query being sent, we use the query inspector, which gives us the string-escaped JSON. This string, we then copy, unescape the ``/"``, copy/paste into MongoDB Compass to generate the step by step aggregation to debug step by step :
> copy the query inspector output
> ![image](https://user-images.githubusercontent.com/110970317/183855633-d5a69124-eef6-4ffb-9495-7e338604c5af.png)
> unescape the quotes
> ![image](https://user-images.githubusercontent.com/110970317/183856173-fcb283d8-d28d-4bf9-a314-4f08afb26b67.png)
> paste into MongoDB Compass
> ![image](https://user-images.githubusercontent.com/110970317/183856289-fee2b7c5-0616-41ac-a994-c8f00c41afcc.png)
> done
> ![image](https://user-images.githubusercontent.com/110970317/183856601-c105689e-c3ec-451d-b824-b2294f53a4ae.png)
> whew!

It would be wonderful to have a similar version to the SQL panels with a query inspector view that gives you the clean query that you can directly copy/paste into a MongoDB Compass to test out:

![image](https://user-images.githubusercontent.com/110970317/183859826-c967bb6e-0920-4b4d-8df5-35b1ccc04201.png)

## 3) Auto indent / Erroneous bracket alerting
Many times, the debugging involves counting brackets, parentheses and curly braces to find where the error could be. Sometimes I copy/paste the unescaped json out of the query inspector (like in the last step) into [jsonlint.com
](jsonlint.com) and let *it* find where my missing brackets are. Again, 30 seconds here, 30 seconds there... it amounts to a bunch of time doing the same repetitive task to debug. It would be great to have ***any*** of the following:
- automatic indentation (or an "Auto-Indent" button somewhere)
- automatic bracket closing (when opening a bracket/parenthesis and pressing enter)
- matching open/closing bracket highlighting when the cursor is on a bracket (like in many IDEs)
- color coding the requests
- JSON lint error highlighting

## 4) Clearer error codes
I know that this might be a longshot since MongoDB doesn't always have the most explicit error messages, but having something like:
![image](https://user-images.githubusercontent.com/110970317/183862886-654881d5-655a-4768-b096-174a795d6d05.png)

instead of

![image](https://user-images.githubusercontent.com/110970317/183862101-e07e41cc-7407-412b-aa4a-25bdf0bcccf4.png)

would **immensely** speed up our debugging efforts when writing complex queries.

### Conclusion

Again, this plugin has really been a breakthrough for us, and is really improving what we can display, but these suggestions would greatly accelerate the time that the data-team spends making our dashboards

Thanks for your consideration!

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
